### PR TITLE
implement CR committee related logic

### PR DESCRIPTION
**Pull request by raindust**

### CODE REVIEW

File: cr/state/state.go [hunk 1]
```diff
@@ -83,8 +101,152 @@ func (s *State) ExistCandidateByNickname(nickname string) bool {
...
+		candidate.cancelHeight = height
+		s.CanceledCandidates[key] = candidate
+		if isPending {
+			delete(s.PendingCandidates, key)
+		} else {
+			delete(s.ActivityCandidates, key)
+		}
+		delete(s.Nicknames, candidate.info.NickName)
+	}, func() {
+		candidate.state = Active
```
Comment (Houshoupei84) [inline cr/state/state.go]: state can be Pending

File: cr/state/state.go [hunk 2]
```diff
@@ -83,8 +101,152 @@ func (s *State) ExistCandidateByNickname(nickname string) bool {
...
+// history to rollback to return error.
+func (s *State) RollbackTo(height uint32) error {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	return s.history.RollbackTo(height)
+}
+
+// FinishVoting will close all voting util next voting period
+func (s *State) FinishVoting() *StateKeyFrame {
+	result := &s.StateKeyFrame
```
Comment (Houshoupei84) [inline cr/state/state.go]: is this need lock?

File: cr/state/committee.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	return &CRMember{
+		Info:             candidate.info,
+		ImpeachmentVotes: 0,
+	}
+}
+
+func (c *Committee) getActiveCRCandidatesDesc() ([]*Candidate, error) {
+	candidates := c.state.GetCandidates(Active)
+	if uint32(len(candidates)) < c.params.CRMemberCount {
+		return nil, errors.New("candidates count less than required count")
```
Comment (Houshoupei84) [inline cr/state/committee.go]: is your mean      count  --->can't ?

File: cr/state/committee.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	} else {
+		if c.store == nil {
+			return errors.New("can't find check point store")
+		}
+		point, err := c.store.GetCheckpoint(height)
+		if err != nil {
+			return err
+		}
+		c.state.StateKeyFrame = point.StateKeyFrame
```
Comment (Houshoupei84) [inline cr/state/committee.go]: may be need lock

File: cr/state/state.go (hunk 1 above)
Comment (raindust) [reply cr/state/state.go]: Fixed

File: cr/state/state.go (hunk 2 above)
Comment (raindust) [reply cr/state/state.go]: Yes, fixed

File: cr/state/committee.go (hunk 4 above)
Comment (raindust) [reply cr/state/committee.go]: Fixed

File: cr/state/committee.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if c.shouldChange(block) {
+		checkpoint := CheckPoint{
+			KeyFrame: c.KeyFrame,
+		}
+
+		if err := c.changeCommitteeMembers(block.Height); err != nil {
+			log.Error("[ProcessBlock] change committee members error: ", err)
```
[the remark this answers is not in the document]
Comment (raindust) [reply cr/state/committee.go]: already handled

File: cr/state/committee.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		}
+
+		checkpoint.StateKeyFrame = *c.state.FinishVoting()
+		if c.store != nil {
+			if err := c.store.SaveCheckpoint(&checkpoint); err != nil {
+				log.Error("[ProcessBlock] save checkpoint error: ", err)
```
[the remark this answers is not in the document]
Comment (raindust) [reply cr/state/committee.go]: already handled

**Pull request merged**